[IRGen] Fix witness table miscompilation.

al45tair opened the pull request:
If a foreign type conforms to a protocol, and that conformance is not resilient, we were emitting an instantantiation function but it would never be called because we weren't emitting the `GenericWitnessTable` that would have contained a reference to it.

This was happening because of a missing `isSynthesizedNonUnique()` call in `isDependentConformance()`.

rdar://97290618

al45tair committed:
[IRGen] Fix witness table miscompilation.

If a foreign type conforms to a protocol, and that conformance is not
resilient, we were emitting an instantantiation function but it would
never be called because we weren't emitting the `GenericWitnessTable`
that would have contained a reference to it.

This was happening because of a missing `isSynthesizedNonUnique()` call
in `isDependentConformance()`.

rdar://97290618

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Hmmm. That exploded unexpectedly. I'll keep digging then.

aschwaighofer commented on lib/IRGen/GenProto.cpp:
```diff
@@ -961,11 +967,13 @@ static bool isDependentConformance(
     if (assocConformance.isInvalid())
       return false;
 
+    auto rootConformance = assocConformance.getConcrete()->getRootConformance();
```
I think you need to guard this by `assocConformance.isAbstract()`

aschwaighofer replied:
... `! assocConformance.isAbstract()`

al45tair replied:
Yes, I figured out as much after doing a local build.

aschwaighofer commented:
Is it possible to add a test case?

al45tair commented:
> Is it possible to add a test case?

Should be, yes.

al45tair commented:
@swift-ci Please smoke test

rjmccall reviewed (changes requested):
LGTM except that it's missing a test case.

al45tair commented:
Having looked at this a bit further, I don't think it's feasible to write a test case that will work on macOS (this only affects synthesized conformances, which means the protocol in question needs to be in the Swift standard library, but on macOS that's dynamically linked and always resilient; building a test case there would require a fairly heroic effort), but it's probably possible on Linux. Trying to get a Linux build up so I can test it.

al45tair committed:
[IRGen] Rearrange fix slightly and add a test case.

isDependentConformance() should always check for synthesized conformances,
so do that further up the function.

Also add a test.

rdar://97290618

al45tair commented:
> Having looked at this a bit further, I don't think it's feasible to write a test case that will work on macOS

It turns out that it _was_ feasible. Just somewhat tricky to get right (the compiler doesn't give very helpful error messages when bits of the `Swift` module are missing; it often just crashes somewhere instead).

al45tair commented:
@swift-ci Please smoke test

aschwaighofer reviewed (approved):
lgtm

al45tair commented:
Failed on Windows and Linux, probably because of the section name. Looking at that now.

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[IRGen][Tests] Fix test case to work for 32-bit architectures.

Changed the test case to pattern match various parts to make it work
on 32-bit.

rdar://97290618

al45tair committed:
[IRGen][Tests] Fix test case to work for Windows.

Windows adds a `comdat` flag.

rdar://97290618

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[IRGen][Tests] Fix test to work on Linux.

Linux uses different section names.

rdar://97690618

al45tair commented:
@swift-ci Please smoke test

rjmccall reviewed (approved):
LGTM
